File: jvm/core/main/scala/com/webank/eggroll/core/containers/meta/DownloadContainersResponse.scala
package com.webank.eggroll.core.containers.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Containers

import scala.collection.JavaConverters._
import scala.language.implicitConversions


case class DownloadContainersResponse(sessionId: String, containerContents: Array[ContainerContent])

object DownloadContainersResponse {
  implicit def deserialize(byteString: ByteString): DownloadContainersResponse = {
    val src = Containers.DownloadContainersResponse.parseFrom(byteString)
    DownloadContainersResponse(
      sessionId = src.getSessionId,
      containerContents = src.getContainerContentList.asScala.map(v =>
        ContainerContent(v.getContainerId, v.getContent.toByteArray, v.getCompressMethod)).toArray
    )
  }

  implicit def serialize(src: DownloadContainersResponse): Array[Byte] = {
    val builder = Containers.DownloadContainersResponse.newBuilder()
      .setSessionId(src.sessionId)
      .addAllContainerContent(src.containerContents.map(ContainerContent.toProto).toList.asJava)
    builder.build().toByteArray
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/pojo/DownloadContainersResponse.java
package com.eggroll.core.pojo;

import com.webank.eggroll.core.meta.Containers;
import lombok.Data;
import org.apache.commons.collections.CollectionUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

@Data
public class DownloadContainersResponse implements RpcMessage {

    Logger log = LoggerFactory.getLogger(DownloadContainersResponse.class);

    private String sessionId;
    private List<ContainerContent> containerContents;


    public DownloadContainersResponse() {

    }

    public DownloadContainersResponse(String sessionId, List<ContainerContent> containerContents) {
        this.sessionId = sessionId;
        this.containerContents = containerContents;
    }

    public Containers.DownloadContainersResponse toProto() {
        Containers.DownloadContainersResponse.Builder builder = Containers.DownloadContainersResponse.newBuilder();
        builder.setSessionId(this.sessionId);

        List<Containers.ContainerContent> list = new ArrayList<>();
        containerContents.forEach(containerContent -> {
            list.add(containerContent.toProto());
        });
        builder.addAllContainerContent(list);
        return builder.build();
    }


    @Override
    public byte[] serialize() {
        return toProto().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Containers.DownloadContainersResponse response = Containers.DownloadContainersResponse.parseFrom(data);
            this.sessionId = response.getSessionId();
            List<Containers.ContainerContent> containerContentList = response.getContainerContentList();
            if (CollectionUtils.isNotEmpty(containerContentList)) {
                List<ContainerContent> containerContentsList = new ArrayList<>();
                containerContentList.forEach(containerContent -> {
                    ContainerContent content = new ContainerContent(containerContent.getContent().toByteArray(),containerContent.getCompressMethod());
                    containerContentsList.add(content);
                });
                this.containerContents = containerContentsList;
            }
        } catch (Exception e) {
            log.error("deserialize error : ", e);
        }
    }
}

